src/lib/integration_test_lib/network_state.ml

```diff
@@ -174,5 +194,49 @@ module Make
                 ; best_tips_by_node = best_tips_by_node'
                 }))
         : _ Event_router.event_subscription ) ;
+    (* handle_breadcrumb_added *)
+    ignore
+      ( Event_router.on event_router Event_type.Breadcrumb_added
+          ~f:(fun node breadcrumb ->
+            update ~f:(fun state ->
+                [%log debug]
+                  "Updating network state with Breadcrumb added to $node"
+                  ~metadata:[ ("node", `String (Node.id node)) ] ;
+                let blocks_seen_by_node' =
+                  let block_set' =
+                    State_hash.Set.add
+                      ( String.Map.find state.blocks_seen_by_node (Node.id node)
+                      |> Option.value ~default:State_hash.Set.empty )
+                      breadcrumb.state_hash
+                  in
+                  String.Map.set state.blocks_seen_by_node ~key:(Node.id node)
```

Comment:
You can write these set operations in a cleaner way. For example, this expression (lines 206-213) could be better written as:

```ocaml
String.Map.update state.blocks_seen_by_node (Node.id node) ~f:(fun block_set ->
  String.Map.add (Option.value block_set ~default:State_hash.Set.empty) breadcrumb.state_hash)
```